qcodes/instrument_drivers/rohde_schwarz/SGS100A.py

```diff
@@ -27,110 +36,220 @@ class RohdeSchwarz_SGS100A(VisaInstrument):
-                           get_parser=self.parse_on_off,
-                           vals=vals.Strings())
-        self.add_parameter('pulsemod_source',
-                           get_cmd='SOUR:PULM:SOUR?',
-                           set_cmd=self.set_pulsemod_source,
-                           vals=vals.Strings())
-        self.add_parameter('ref_osc_source',
-                           label='Reference oscillator source',
-                           get_cmd='SOUR:ROSC:SOUR?',
-                           set_cmd='SOUR:ROSC:SOUR {}',
-                           vals=vals.Enum('INT', 'EXT'))
-        # Frequency mw_source outputs when used as a reference
-        self.add_parameter('ref_osc_output_freq',
-                           label='Reference oscillator output frequency',
-                           get_cmd='SOUR:ROSC:OUTP:FREQ?',
-                           set_cmd='SOUR:ROSC:OUTP:FREQ {}',
-                           vals=vals.Enum('10MHz', '100MHz', '1000MHz'))
-        # Frequency of the external reference mw_source uses
-        self.add_parameter('ref_osc_external_freq',
-                           label='Reference oscillator external frequency',
-                           get_cmd='SOUR:ROSC:EXT:FREQ?',
-                           set_cmd='SOUR:ROSC:EXT:FREQ {}',
-                           vals=vals.Enum('10MHz', '100MHz', '1000MHz'))
-
-        self.add_function('reset', call_cmd='*RST')
-        self.add_function('run_self_tests', call_cmd='*TST?')
+    def __init__(self, name, address, max_frequency=12.75e9, **kwargs):
+        super().__init__(name, address, terminator="\n", **kwargs)
+
+        self.add_parameter(
+            name="frequency",
+            label="Frequency",
+            unit="Hz",
+            get_cmd="SOUR:FREQ?",
+            set_cmd="SOUR:FREQ {:.2f}",
+            get_parser=float,
+            vals=vals.Numbers(1e6, max_frequency)
+        )
+        self.add_parameter(
+            name="phase",
+            label="Phase",
+            unit="deg",
+            get_cmd="SOUR:PHAS?",
+            set_cmd="SOUR:PHAS {:.2f}",
+            get_parser=float,
+            vals=vals.Numbers(0, 360)
+        )
+        self.add_parameter(
+            name="power",
+            label="Power",
+            unit="dBm",
+            get_cmd="SOUR:POW?",
+            set_cmd="SOUR:POW {:.2f}",
+            get_parser=float,
+            vals=vals.Numbers(-120, 25)
+        )
+        self.add_parameter(
+            name='maximum_power',
+            label='Maximum power',
+            unit='dBm',
+            get_cmd=lambda: self.power.vals._max_value,
+            set_cmd=lambda power: setattr(self.power.vals, '_max_value', power),
+            initial_value=self.power.vals._max_value,
+            # VISA commands do not seem to influence max power
+            # get_cmd='SOURce:POWer:LIMit?',
+            # set_cmd='SOURce:POWer:LIMit:AMPLitude {:.2f}',
```

Comment:
The manual says:

> Sets the level limit.
> 
> The value specifies the upper limit of the level at the [RF Out] connector. A message appears if an attempt is made to set a level above this limit and the level at the RF output is confined to the upper limit. However, the **level indication is not influenced**.

So maybe it's working and raising an error you're not detecting.

Reply:
Good catch! I was wondering why the Rabi frequencies weren't increasing past 0 dBm, forgot I set it to 0 dBm a long long time ago. You get a medal🥇